src/ID3Helpers.js

```diff
@@ -96,55 +96,61 @@ module.exports.getTagsFromBuffer = function(filebuffer, options) {
     }
 
-    let identifierSize = 4
-    let textframeHeaderSize = 10
-    if(ID3Version === 2) {
-        identifierSize = 3
-        textframeHeaderSize = 6
-    }
+    const frameIdentifierSize = (ID3Version === 2) ? 3 : 4
+    const frameHeaderSize = (ID3Version === 2) ? 6 : 10
 
-    while(currentPosition < ID3FrameBody.length && ID3FrameBody[currentPosition] !== 0x00) {
-        const bodyFrameHeader = Buffer.alloc(textframeHeaderSize)
-        ID3FrameBody.copy(bodyFrameHeader, 0, currentPosition)
+    while(currentPosition < ID3TagBody.length && ID3TagBody[currentPosition] !== 0x00) {
+        const frameHeader = Buffer.alloc(frameHeaderSize)
+        ID3TagBody.copy(frameHeader, 0, currentPosition)
 
-        let decodeSize = false
-        if(ID3Version === 4) {
-            decodeSize = true
-        }
-        let bodyFrameSize = ID3Util.getFrameSize(bodyFrameHeader, decodeSize, ID3Version)
-        if(bodyFrameSize + 10 > (ID3FrameBody.length - currentPosition)) {
-            break
-        }
-        const specName = bodyFrameHeader.toString('utf8', 0, identifierSize)
-        if(options.exclude instanceof Array && options.exclude.includes(specName) || options.include instanceof Array && !options.include.includes(specName)) {
-            currentPosition += bodyFrameSize + textframeHeaderSize
+        const frameIdentifier = frameHeader.toString('utf8', 0, frameIdentifierSize)
+        const decodeSize = ID3Version === 4
+        const frameBodySize = ID3Util.getFrameSize(frameHeader, decodeSize, ID3Version)
+        // It's possible to discard frames via options.exclude/options.include
+        // If that is the case, skip this frame and continue with the next
+        if(isFrameDiscardedByOptions(frameIdentifier, options)) {
+            currentPosition += frameBodySize + frameHeaderSize
             continue
         }
-        const frameHeaderFlags = ID3Util.parseFrameHeaderFlags(bodyFrameHeader, ID3Version)
-        if(frameHeaderFlags.dataLengthIndicator) {
-            bodyFrameSize -= 4
+        // Prevent errors when the current frame's size exceeds the remaining tags size (e.g. due to broken size bytes).
+        if(frameBodySize + frameHeaderSize > (ID3TagBody.length - currentPosition)) {
+            break
         }
-        const bodyFrameBuffer = Buffer.alloc(bodyFrameSize)
-        ID3FrameBody.copy(bodyFrameBuffer, 0, currentPosition + textframeHeaderSize + (frameHeaderFlags.dataLengthIndicator ? 4 : 0))
+
+        const frameHeaderFlags = ID3Util.parseFrameHeaderFlags(frameHeader, ID3Version)
+        // Frames may have a 32-bit data length indicator appended after their header,
+        // if that is the case, the real body starts after those 4 bytes.
+        const frameBodyOffset = frameHeaderFlags.dataLengthIndicator ? 4 : 0
+        const bodyFrameBuffer = Buffer.alloc(frameBodySize - frameBodyOffset)
+        ID3TagBody.copy(bodyFrameBuffer, 0, currentPosition + frameHeaderSize + frameBodyOffset)
+
```

Comment:
Do we need a copy? If we do read only operations, we could use a view instead with `Buffer.slice`.
https://nodejs.org/docs/latest-v10.x/api/buffer.html#buffer_buf_slice_start_end
```js
const frameBody = ID3TagBody.slice(currentPosition + frameHeaderSize + frameBodyOffset)
```

Reply:
I don't think we need to copy but I wasn't sure if read-only is guaranteed throughout the whole code so I left it in for now